src/yz_index.erl

```diff
@@ -220,33 +212,81 @@ local_create(Name) ->
+core_create(Name, SchemaName, CoreProps) ->
+    %% Solr won't allow us to CREATE a new core when one with the same
+    %% properties already exists. This is problematic when treating
+    %% create as an idempotent operation, and moreso when creation fails
+    %% for one reason or another, or AAE believes the index needs to be
+    %% re-created (i.e. bad schema, missing core.properties, etc...).
+    %% Without first unloading the core Solr may believe the core still
+    %% exists (i.e. exists in memory, but there's no core.properties on
+    %% disk), but AAE may disagree and enter a cycle of continuously
+    %% trying to CREATE the core, making it impossible for AAE to
+    %% recover Solr from these states.
+    %%
+    %% To address this, we first unload (local_remove) the core from Solr.
+    %% This may be a moderately expensive process, but given we should
+    %% only be issuing creation requests when either AAE believes Solr is
+    %% missing the index, or we are creating it for the first time, the
+    %% impact should be minimal (and somewhat unavoidable; if AAE is
+    %% replacing the whole index, that's what it's going to do anyway).
+    ok = local_remove(Name, [{core, Name}]),
+    case yz_solr:core(create, CoreProps) of
+        {ok, _, _} ->
+            lager:info("Created index ~s with schema ~s",
+                       [Name, SchemaName]),
+            ok;
+        {error, exists} ->
+            lager:info("Index ~s already exists in Solr, "
+                       "but not in Riak metadata",
+                       [Name]);
+        {error, Err} ->
+            lager:error("Couldn't create index ~s: ~p", [Name, Err])
+    end.
+
 %% @doc Remove the index `Name' locally.
 -spec local_remove(index_name()) -> ok.
 local_remove(Name) ->
     CoreProps = [
-                    {core, Name},
-                    {delete_instance, "true"}
+                 {core, Name},
+                 {delete_instance, "true"}
                 ],
-    {ok, _, _} = yz_solr:core(remove, CoreProps),
+    local_remove(Name, CoreProps).
+
+-spec local_remove(index_name(), [{core | delete_instance | delete_index | delete_data_dir,
+                                   index_name() | string()}]) -> ok.
+local_remove(Name, CoreProps) ->
+    case yz_solr:core(remove, CoreProps) of
+        {ok, _, _} ->
+            lager:info("Unloaded previous instance of index ~s", [Name]);
+        {error, {ok, "400", _, Resp}} ->
+            lager:info("Couldn't unload index ~s prior to creating "
+                       "a new instance of it. This is likely the first "
+                       "time this index has been created. Solr "
+                       "responded with ~s.", [Name, Resp]);
+        {error, UnloadErr} ->
+            lager:error("Couldn't unload index ~s prior to creating "
+                        "a new instance of it: ~p", [Name, UnloadErr])
+    end,
```

Comment:
I think all clauses here return `ok`, you should be fine without the extra return value.